Assassination - article quick fix

zonkmachine opened the pull request:
In the Assassination module the answer to chat question "Where do I find {target}" is always the same and the ending looks a bit off as the string doesn't cater for the ship model, a/an.

`... The ship is Wave and has registration ID QA-7856.`

The quick solution is to assume article 'a' as by chance it fits in with all ship models except AC-33. Now the string is off only one out of 27 times.

impaktor commented:
Would "of model" fix it?

"of model AC..."
"of model Wave"

or "of make/", "type"?

bszlrd commented:
Out of scope, but how about having a "a" or "an" defined in the ship.jsons, so missions could see which to use? Or I'm overthinking again?

zonkmachine commented:
> Would "of model" fix it?
> 
> "of model AC..." "of model Wave"
> 
> or "of make/", "type"?

You would make it grammatically correct but it's not how people speak. It would still be annoying.


> Out of scope, but how about having a "a" or "an" defined in the ship.jsons, so missions could see which to use? Or I'm overthinking again?

I think you would need to make the ships translatable somehow but I don't know if that would work in a json file. Maybe have ship terms defined in a lua file, make, synonyms, etc. and that could be easily translatable.

zonkmachine commented:
OK, I'll try a simple "ships" json file.

zonkmachine commented:
This now works but only with new games so it wouldn't be for the next release. I think maybe, especially if the AC-33 isn't even ever selected for assassination targets, that the first one-liner still has the merits of just bringing good and you can add it to the next release. The new fuller version is better in the long run and can be applied to other modules as well.

zonkmachine commented:
**"a Pumpkinseed Police"**

I guess this wouldn't look good in the game.

**The ship is a Pumpkinseed Police and has registration id {shipregid}.**

Maybe

**"a Pumpkinseed police ship"** ?
**The ship is a Pumpkinseed police ship and has registration id {shipregid}.**

zonkmachine commented:
![assassino](https://github.com/user-attachments/assets/1d719072-2069-476e-a14f-5edfe4962867)

zonkmachine commented:
> I'm not convinced the best and simplest work-around is to re-word the context in which it is used, something along the lines of my initial suggestion

Unfortunately that creates a language that isn't flowing very well and that's how it's done all over the code. I've tried rephrasing stuff like that and it doesn't work as well as the real deal. On the other hand, I've poked around ways to address it and in precisely every case it runs away in complexity super fast. My wet dream right now is to just pipe the strings, at least some of them, to an external app that can fix them up on the fly with at least the most basic stuff, like a/an or genitive s or whatever is most worth it in other languages.

I think this way of doing it here, by making the ships translatable, is a reasonable way of doing it. Lot's of strings but a simple to understand idea so it should be easy on the translators. It will take some consideration though. You could add more fields for the ships so some of them could have alternative names or nicknames. The AC-33 could be either AC-33 or Dropstar or both.

I think it is also reasonable to merge the original one-line fix for the up coming stable bug fix. It's a small change but that string is on ever single assassination mission and it looks bad.

I approve of this message.

zonkmachine commented:
> ```json
>     "ac33": {
>         "description": "",
>         "message": "an AC-33 Dropstar"
>     },
> ```

As I understand it, as these are new translations, the strings here would be pushed to all other languages and come out with English articles. That wouldn't be good. Is there a way to do a first push to Transifex with ships.json without articles so the translations will look the same as today and then change the strings in a later commit?

zonkmachine commented:
> push to Transifex with ships.json without articles

Do all languages at the same time and with the same English translation without article or with a rough automatic translation? This is where we could use Grok right?

impaktor commented:
> Is there a way to do a first push to Transifex with ships.json without articles so the translations will look the same as today and then change the strings in a later commit?

No, not that I know of.

> Do all languages at the same time and with the same English translation without article or with a rough automatic translation? This is where we could use Grok right?

Hm? When you push they get the default English until translated. However, the cron-job that pulls the strings back to master doesn't like to run, so at the moment nothing gets back to master until I kick the machine (which we also do when releasing new version)

zonkmachine commented:
> When you push they get the default English until translated.

Yes, but the strings are still in the English original. If we merge this, all languages will show ups with the English article. I looked at some language files and it looks like Swedish, Danish, Italian, and French, all could use their ship file with just the current name added. You don't need to be a translator to manage that. If we pushed these file pretranslated, would Transifex accept that?

zonkmachine commented:
OK. Definitely not a quick fix no longer.

Added strings for Search and Rescue too. The variable names are not unison for the modules. Will fix.

impaktor commented:
Would like thoughts from @sturnclaw here, on if adding a translation module ship is the best solution (and if it is to be added, would the name be "ships-core" or "ships"?)
 
(Besides, this will need further hooking up into transifex upon merge, by me)

zonkmachine commented:
> (Besides, this will need further hooking up into transifex upon merge, by me)

It's a bit like #5852. Did you have to do anything special on merge?

impaktor commented:
@zonkmachine no, difference is this adds a _new_ module, that needs some configuring on transifex side.

Mc-Pain commented:
> Where do I find {target}
< [...] The target is using Wave with registration ID QA-7856.

zonkmachine commented:
> Where do I find {target}
> < [...] The target is using Wave with registration ID QA-7856.

Please be more verbose. What am I looking at here?

Mc-Pain commented:
> > Where do I find {target}
> > < [...] The target is using Wave with registration ID QA-7856.
> 
> Please be more verbose. What am I looking at here?

Oops. It was not supposed to be a quote.
It was a suggestion to change answer in order to resolve the issue.

zonkmachine commented:
> Oops. It was not supposed to be a quote.
> It was a suggestion to change answer in order to resolve the issue.

Sorry, English is not my first language. ; )
A - What is that issue?
B - What is the solution?

zonkmachine commented:
> < [...] The target is using Wave with registration ID QA-7856.

I think you still need an article in there for this sentence to be complete.
 **The target is using a Wave with registration ID QA-7856.**

zonkmachine commented:
Rebased on master and squashed. I don't know how to deal with other languages. I can add in all of them with just the base name like in the Swedish ship name file or just leave it to the translators to solve. It may be easier to add in the files via this PR or the untranslated ship names will pop up with English articles where they should not.

mwerle commented on data/modules/Assassination/Assassination.lua:
```diff
@@ -216,7 +217,7 @@ local makeAdvert = function (station)
 		threat = threat,
 		shipid = hullConfig.id,
-		shipname = ShipDef[hullConfig.id].name,
+		shipname = ShipDef[hullConfig.id].modelName,
```
It might be better to rename the variable to indicate to other developers that this is no longer the actual shipname, but the modelname which must then be used as a key to the translation.

Alternatively, why not perform the translation here which means you don't have to call the translation function everywhere else in this module?

mwerle commented on data/modules/SearchRescue/SearchRescue.lua:
```diff
@@ -727,7 +728,8 @@ local onChat = function (form, ref, option)
 			problem      = ad.problem,
 			cash         = Format.Money(ad.reward),
-			ship         = shipdef.name,
+			shipname     = shipdef.name,				-- Ship name, base form ex. 'Malabar'
```
This name is not translated.

For example, german:
`shipname` == "Sinonatrix (Police)"
`shipsing` == "ein Sinonatrix Polizeischiff"

Even in english there will be a difference, so somewhat confusing and disjointed player experience I think.

mwerle commented on data/modules/SearchRescue/SearchRescue.lua:
```diff
@@ -2120,7 +2122,7 @@ local buildMissionDescription = function(mission)
 	end
 
-	local shipname = ShipDef[mission.shipid].name
+	local shipname = ShipDef[mission.shipid].name -- local shipname = ls[ShipDef[mission.shipid].modelName]
```
... what is the meaning of this line?

zonkmachine replied:
Looks like commented out experimental code.

mwerle reviewed (changes requested):
Thank you for your work, but after looking at this PR I think it is only the start of translating ship names.

Firstly, it only adds translations for "singular" ship names.

Secondly it only updates the Assassination and SearchRescue missions. There are many other areas in the game which display the ship name to the user. While most of the ship names are literally just names, a few contain a descriptor as well, which needs translating.

Furthermore some of the ship names are real nouns/descriptors so it might actually make sense for different languages to provide a different name. For example "Wave"(en) -> "Welle"(de).

zonkmachine commented on data/lang/ships/sv.json:
The Swedish translation file is part of a discussion on how to deal with the other languages in this PR. If the file above is duplicated for all languages the ship names will show up like they do today instead of suddenly showing up anglified.

zonkmachine commented:
> Thank you for your work, but after looking at this PR I think it is only the start of translating ship names.
> 
> Firstly, it only adds translations for "singular" ship names.
> 
> Secondly it only updates the Assassination and SearchRescue missions. There are many other areas in the game which display the ship name to the user. While most of the ship names are literally just names, a few contain a descriptor as well, which needs translating.

Yes, it's an early experiment. I don't know if it's a good idea to accommodate 'plural' in the code but I'm sure going to poke into it. If needed it could perhaps be handled by dedicated strings where a known model is used.

> Furthermore some of the ship names are real nouns/descriptors so it might actually make sense for different languages to provide a different name. For example "Wave"(en) -> "Welle"(de).

Good point. Thanks for the feedback. I'll look into the rest of your comments in a, hopefully, short while.
The branch has been rebased on master.

zonkmachine commented on data/lang/module-searchrescue/en.json:
```diff
@@ -396,8 +396,8 @@
   "FLAVOUR_7_INTROTEXT": {
-    "description": "",
-    "message": "My name is {name} of {entity}. Our only ship, {ship} {shiplabel}, is drifting with a crew of {crew} after {problem}. I am willing to pay {cash} for immediate help."
+    "description": "{shipsing} is translated in /data/lang/ships/",
```
Now it follows basic grammar NAME_BASE, NAME_INDEF and NAME_DEF. Currently the tokens are in upper case to follow the rest of the lang code.

zonkmachine commented:
Q - I changed the ship tokens to upper case to harmonize with the rest of the translation files. It has some issues as now I need to go string.upper on the ship names. You could instead change the ship names to be an upper case token? I'm inclined to go back to the original lower case and not touch the ship json files at this point. Any ideas on this? Should I go all lower case? Is there a convention for this?

sturnclaw commented:
For the record, since I've not chimed in on this yet - I very much appreciate the initiative taken here! We've definitely needed better ways to handle string interpolation of various names, and I think the definite, indefinite, generic approach works well. You might want an additional form for plural, though I do not know if it would be helpful or necessary in most languages; e.g. in English we'd say "the targets fly three Wave-class ships" or (informal) "they're flying three Waves".

sturnclaw commented:
A third thing to consider: in English, modifying adjectives and nouns can be inserted between the article ('a', 'the') and the actual noun ('Coronatrix') without changing the spelling of either. Does this also keep with Swedish, etc. or would mission modules need to provide custom translated ship names if they wanted to include adjectives (i.e. "the target flies a militarized AC33")? If the former works well, it may suffice to have the definite/indefinite forms omit the article, and have the article in the language string the ship name is to be interpolated into.

zonkmachine commented:
> in English, modifying adjectives and nouns can be inserted between the article ('a', 'the') and the actual noun ('Coronatrix') without changing the spelling of either

Give me some examples. I thought of modifying nouns just an hour ago and my understanding is that it will exponentially increase the difficulty of translation. In English (this is my understanding), a/an is decided by the pronunciation of the following word.

A Coronatrix
An old Coronatrix

In Swedish the article depends on the noun and the adjective is changed too. Sometimes...
En Coronatrix
En gammal Coronatrix

But for the word ship:
Ett skepp
Ett gammalt skepp 

So, old in english didn't change, but in swedish it did. And this is only two languages. Spontaneously I think this will be hard. I think rather quickly we would need to use something like a dedicated translation software (Appertium or something similar) where you can add grammatical rules between two languages to shuffle the order between them and handle suffixes etc. It would probably be better to look into some dedicated software for this.
I haven't even thought much about plural.

I think this would need a bigger discussion involving more than two languages. Maybe it's easier than I think. Like how we only could have renamed AC-33 to BC-33 and then all ships in English would have been "a {ship}".

zonkmachine commented on data/ships/ac33.json:
```diff
@@ -1,6 +1,7 @@
 {
 	"model": "ac33",
 	"name": "AC-33 Dropstar",
+	"i18n_key": "AC33",
```
* @sturnclaw Is this what you had in mind?
* What do I do with the missiles. Should they also have the i18n_key tags?

zonkmachine replied:
> @sturnclaw Is this what you had in mind?

Looks like I need a common component like SHIP.

```json
"i18n_key": "SHIP_AC33",
```

zonkmachine replied:
Fixed. I've looked at how it's done with the hardpoints and took it from there. Just the keys are added for now.

zonkmachine commented:
@sturnclaw Sorry about the noise. I managed to get the i18n_key to work. I did it from the c++ side though. Is this correct?

The code, at least the transifex stuff, needs commenting.

zonkmachine commented:
OK, We're rebased and squashed to two commits for easier review. It now must be tested together with #6241

zonkmachine commented:
Added translatable ship names in the Views.
* Person Info - Ship Information (model spinner)
* Ship Market - Ship list and model spinner
* Equipment Market (model spinner)

zonkmachine commented:
I prematurely removed the ship i18n tags ahead of #6241. Added back for the time being or the PR will break when testing.

zonkmachine commented on data/modules/Assassination/Assassination.lua:
```diff
@@ -490,7 +490,7 @@ local function buildMissionDescription(mission)
 		{ l.DISTANCE, dist.." "..lc.UNIT_LY },
 		false,
-		{ l.SHIP, mission.shipname },
+		{ l.SHIP, ls[ShipDef[mission.shipid].i18n_key .. "_BASE"] },
```
Maybe it's enough with the bare i18n_key as **_BASE**? Just loose that suffix?

zonkmachine replied:
> it's probably best to split `ShipDef[mission.shipid]` into its own statement

2b731d4d29e394e999e5eb0b852a4895ce6a8446

zonkmachine commented:
Rebased on master and commit message improved

sturnclaw commented on src/lua/LuaShipDef.cpp:
```diff
@@ -21,6 +21,29 @@
+ * > local ship = ShipDef[ad.shipid].i18n_key            -- 'NATRIX'
+ * >
+ * > local ship_def = ls[ship .. "_DEF"],                -- 'NATRIX_DEF' - 'the Natrix'
+ * > print("We're counting on " .. ship_def .. " to give us some resistance!)
```
Nitpick-iest of nitpicks: missing closing quotation mark here. I only mention it because someone new to programming may try to copy-paste from the docs while writing a module.

sturnclaw reviewed (approved):
> @sturnclaw I kept the esoteric stuff separate ([77bf4de](https://github.com/pioneerspacesim/pioneer/commit/77bf4de60f905cebb9d5732b33675766d07b6adc)). Esoteric meaning C++. The meaty stuff, (lua) I think I have some kind of grasp on but src/lua/_stuff_, not so much. I'll let this one rest now, no more force pushes.

Looks good to me, adding a single field to LuaShipDef is very straightforward and doesn't need much rocket surgery.

I like what I'm seeing here, though I don't see any (current) consumers of the `SHIP_DEF` form? On the whole, this PR rectifies a long-running deficiency in the localization capability of the game, and you get two big thumbs up for improving that!
